### ISSUE TITLE
com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl is occupying more heapMemory and finally resulting in OOM Error

### ISSUE THREAD

**Issue by RameshP96** [hazelcast/hazelcast#14131]
Hi Peter, Matko,

we are experiencing heavy Memory consumption problem by `com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl`, occupying almost more than half of the heapMemory after running for few hours/days, and finally process goes `OutOfMemoryError`.
We are using Hazelcast 3.9.1 in our product. 

Initially product was using 2.6.9 and we upgraded it to 3.8.1 in May 2017 and again to 3.9.1 in Dec 2017.

After running for few hours and analyzing the heapdump, following is the information given by java profiler (Eclipse Mat tool) with hazelcast3.9.1, with cluster-size of 2 members

```
One instance of "com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl" loaded by "sun.misc.Launcher$AppClassLoader @ 0x1348e440" occupies 59,510,416 (40.10%) bytes. The memory is accumulated in one instance of "com.hazelcast.spi.impl.eventservice.impl.EventServiceSegment" loaded by "sun.misc.Launcher$AppClassLoader @ 0x1348e440".
```

Can you please let us know what we should check for in the `EventServiceSegment` object? I have attached screen-capture of this class from a java-profiler by name "EventServiceSegment.png", please let me know what i need to look at in this object? or if you need anymore from me.

![eventservicesegment](https://user-images.githubusercontent.com/44966026/48559265-48ce5e00-e911-11e8-9d6d-9e5b63a45ac3.png)

![eventservicesegment-retainedby](https://user-images.githubusercontent.com/44966026/48559277-4e2ba880-e911-11e8-8d8e-f09236b7c642.png)

We are using listeners in the product (use addMessageListener(...)/removeMessageListener(...)) on topics. 

I initially updated an already closed issue (https://github.com/hazelcast/hazelcast/issues/13005) which looks similar to what I am experiencing in our product, but thought it is better i open a new issue so that I can provide any context related to the problem that we are experiencing.

-Dhazelcast.icmp.enabled=true
Java - JDK8u152
Happens in Linux/Windows

Some more log messages...
------------------------------
```
INFO #AddressPicker                      # [LOCAL] [YUAJCjsFSN] [3.9.1] Picking address configured by property 'hazelcast.local.localAddress'
INFO #AddressPicker                      # [LOCAL] [YUAJCjsFSN] [3.9.1] Picked [MyTestBox]:7004, using socket ServerSocket[addr=/0.0.0.0,localport=7004], bind any local is true
INFO #AddressPicker                      # [LOCAL] [YUAJCjsFSN] [3.9.1] Using public address: [MyTestBox]:7004
INFO #system                             # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Hazelcast 3.9.1 (20171130 - feca534) starting at [MyTestBox]:7004
INFO #system                             # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
INFO #system                             # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Configured Hazelcast Serialization version: 1
INFO #BackpressureRegulator              # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Backpressure is disabled
INFO #Node                               # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Creating TcpIpJoiner
INFO #OperationExecutorImpl              # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Starting 4 partition threads and 3 generic threads (1 dedicated for priority tasks)
INFO #Diagnostics                        # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
INFO #LifecycleService                   # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] [MyTestBox]:7004 is STARTING
INFO #system                             # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Cluster version set to 3.9
INFO #ClusterService                     # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] 


INFO   | jvm 1    | 2018/11/13 20:19:45 |        5 | @tomcat32-resource <WrapperStartStopAppMain> []  WARN #PhoneHome                          # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] hazelcast.version.check.enabled property is deprecated. Please use hazelcast.phone.home.enabled instead to disable phone home.
INFO   | jvm 1    | 2018/11/13 20:19:45 |        5 | @tomcat32-resource <WrapperStartStopAppMain> []  INFO #ManagementService                  # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] Hazelcast JMX agent enabled.
INFO   | jvm 1    | 2018/11/13 20:19:45 |        5 | @tomcat32-resource <WrapperStartStopAppMain> []  INFO #LifecycleService                   # [MyTestBox]:7004 [YUAJCjsFSN] [3.9.1] [MyTestBox]:7004 is STARTED
```

Thanks and regards
Ramesh

**Comment by RameshP96** [comment]
Hi Matko, Peter,

Can you please help on this, what could be the reason for the problem so that I can look from my side in the heapdump? please consider this as a high priority issue as we have severity issue opened by our customers that they have to restart our product every few days.

Following is the latest heapdump related information, The com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl is occupying 56% of the heapMemory, profiler says following is the biggest object (taken 247 MB), please refer below screen
![biggestobjects_dominators_21nov](https://user-images.githubusercontent.com/44966026/48847643-149ee580-edc8-11e8-920d-7a96fc96333c.png)

The ObjectExplorer of com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl , this shows that com.hazelcast.spi.impl.eventservice.impl.EventServiceSegment object inside 'segments' HashMap is taking total 247MB, please refer below screen
![eventserviceimpl_segments_eventservicesegment_nov21](https://user-images.githubusercontent.com/44966026/48847684-26808880-edc8-11e8-90f8-940b43cb66cd.png)

When i explored com.hazelcast.spi.impl.eventservice.impl.EventServiceSegment object (occupying 247Mb), i see 'registrations' and 'registrationIdMap' ConcurrentHashMap objects in it are occupying  around 36Mb (17.9Mb, 17.9Mb) but not sure about the rest of the 211 Mb heap of EventServiceSegment object? please refer below screen
![eventservicesegment_nov21](https://user-images.githubusercontent.com/44966026/48848104-28971700-edc9-11e8-9815-4cf27076f90f.png)

can you please help me identify what can take the 211Mb heap?, need to know how to identify/fix this, we have escalation going on with our customers.
To optimize on the 36Mb (17.9 + 17.9) occupied in EventServiceSegment, when i looked at registrations ConcurrentHashMap, it has ConcurrentHashMap object created per topic which is understandable, but I am not sure what are the 2,87,481 ConcurrentHashMap ("Registration --> true") object with in "serversChannel" topic corresponds to?
![eventservicesegment_serverschannel-topic_childhashmaps_nov21](https://user-images.githubusercontent.com/44966026/48848998-6c8b1b80-edcb-11e8-9b14-00c1e9dccbd3.png)
 
I quickly checked in my DEV environment where similar workload runs, 
- I kept "System.out"/SOP statements at 6 places (with in singleton objects) where we do "serverChannel(topic object)".addMessageListener(). other than those 6, i don't see anywhere else we do MessageListener for "serverChannel" topic.
- Also kept "System.out"/SOP statements from where we publish messages to "serverChannel" topic (it is again one place).
But haven't seen any SOP statements getting printed in logs other than the 6 places where we add MessagesListeners.

I will try deploy these SOP statements in myTEST box where i could reproduce this issue and see whether I see too-many message-listeners are getting added other than the 6 places?

Can you please help me identify the root-cause for this? if anything required to be fixed at my end I am more than happy to handle that.

Do you think this problem is to do with application-code or it is related to hazelcast library?
Do you think using 3.11 will fix this problem ?
I see that this issue is added/marked for "3.12 milestone", does this mean your team is already identified similar issue in hazelcast and it is being worked currently for 3.12 milestone?

Thanks and regards
Ramesh

**Comment by RameshP96** [comment]
Hi Matko, Peter,

Any update for me on this?

Thanks and regards
Ramesh

**Comment by RameshP96** [comment]
Hi Matko, Peter,

Any update for me on this?

Thanks and regards
Ramesh

**Comment by mmedenjak** [comment]
Hi @RameshP96 !

sorry for the late reply. I would assume there is a leak of listener registrations somewhere. Are you cleaning up and deregistering listeners as well?
Can you send us the heap dump so I can take a look?

**Comment by RameshP96** [comment]
Hi Matko,
Thanks for you reply, saw your update today only (have been busy with my other stuff).

In fact I have resolved this issue a month ago and delivered a patch with our product changes.
You are right the primary problem was that the listener registrations were piling up for the "serversChannel" topic which was the root cause.

For others whoever encounters similar problem in future:
- Specifically in our case we have a spring-bean which registers listener to hazel-cast for specific event (in our case "serversChannel" topic). This listener code was written by someone long back in our product but it never gave any problem until product was using hazel-case 2.6.9 version but when we recently upgraded product to use 3.9.1, we started seeing this listener-registration leak.

I was in an assumption that the spring-bean which does hazelcast-listener registration/deregistration in our product is a singleton, but when debugging understood that it is at different scope so the class was getting created every "certain frequency", does hazelcast-listener registration, does intended work and finally while in its @PreDestroy method it deregisters the hazelcast-listener.

I see that with hazelcast-3.9.1 (atleast) the listener objects are not getting removed though product calls deregistration, which to me seems hazelcast problem.

But i workaround that by **not "doing registering to hazelcast-listener from the non-singleton bean**" instead "**did the hazel-cast listener registration from a singleton-bean**" which does single-time the listener-registration (does deregistration only when product is going shutdown) so that registration/deregistration to hazelcast is minimized to single-time to avoid memory growth due to hazelcast-registrations
I have used Observer pattern (between my non-singleton-bean and singleton-bean) such that my non-singleton-bean gets notified when singleton-bean gets hazelcast-events.

Matko, to reproduce the problem, may be we can use spring @PostConstruct to do hazelcast-registration and @PreDestroy method to deregister the listener and can schedule the bean to run every 5 seconds i am sure we should see this hazelcast-listener-registration objects getting pilled up. 

Best regards
Ramesh
